core/shared/src/main/scala/scalaz/zio/ZManaged.scala

```diff
@@ -73,6 +73,48 @@ final case class ZManaged[-R, +E, +A](reserve: ZIO[R, E, Reservation[R, E, A]])
       }
     }
 
+  final def foldM[R1 <: R, E2, B](
+    failure: E => ZManaged[R1, E2, B],
+    success: A => ZManaged[R1, E2, B]
+  ): ZManaged[R1, E2, B] =
+    ZManaged[R1, E2, B] {
+      Ref.make[ZIO[R1, Nothing, Any]](IO.unit).map { finalizers =>
+        Reservation(
+          acquire = {
+            val direct =
+              self.reserve
+                .flatMap(res => finalizers.update(fs => res.release *> fs).const(res))
+                .uninterruptible
```

Comment:
I have a feeling that all of these `uninterrruptible` regions are _overly_ aggressive; and rather, `uninterruptibleMask` should be used to limit the extent of the uninterruptible regions.

Reply:
indeed but that's just because ZManaged itself has not been updated to support the fancy new operators :-)

Reply:
I'll take a look